vuln-reach/src/javascript/lang/symbol_table.rs

```diff
@@ -341,19 +415,33 @@ impl<'a> SymbolTable<'a> {
         // layers until a scope with the node name is declared.
         let (decl_scope, decl_node) =
             loop {
+                // Retrieve the scope.
                 let scope = &self.scopes[parent_scope_index];
+
+                // If the scope contains a declaration named like the node we are looking up,
+                // return the scope and the declaration node.
                 if let Some(decl_node) = scope.names.iter().find_map(|&node| {
                     if self.tree.repr_of(node) == name { Some(node) } else { None }
                 }) {
                     break (scope, decl_node);
                 }
 
+                // If we walked all the way to the program node, we haven't found the
+                // declaration. We can directly return `None` from here.
                 if parent_scope_index == 0 {
                     return None;
                 }
 
                 let cur_level = scope.level;
+
                 // Walking backwards, find the first scope with level less than the current.
+                //
+                // This assumes that:
+                // - all the sibling scopes have the same level as cur_level,
+                // - all scopes appear to the right of their parent scope in self.scopes
+                // - all scopes' siblings appear contiguously
+                // - no scopes of level less than or equal to the parent's level appear between
+                //   a parent scope and its child scopes
```

Comment:
It seems weird to have these assumptions documented here. Shouldn't this be part of an API for the scopes maybe? Otherwise how would someone changing the actual data being accessed know about the invariants they need to uphold?

Reply:
I'm torn on this. These invariants flow from the fact that the scopes are accessed in a depth-first way. I removed this for the time being, I'll think of something.